# 00. Contents

### 01. Importing libraries
### 02. Importing data frames
### 03. Finding the mean for order numbers
### 04. Creating loyalty flag
### 05. Identifying spending habits by customer loyalty
### 06. Targeting high spenders
### 07. Identifying the frequent customers

# 01. Importing libraries

In [2]:
import pandas as pd
import numpy as np
import os

# 02. Importing dataframes

In [6]:
# Setting up a path
path = r'/Users/agne/Documents/Studies/Data Analysis/Study Materials/Python/Instacart Basket Analysis 2020 11'

In [9]:
ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_new_var.pkl'))

# 03. Finding the mean for order numbers

In [16]:
# Grouping and aggregating the info based on department ID
ords_prods.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

The limited dataframe (with 1000000 rows) is not much different from the one that contains the whole information. There are some inaccuracies (such as, department 21 having a mean of 21.9 in the limited dataframe and 22.9 in the full one), but the margin of difference is only around 1.

# 04. Creating loyalty flag

In [29]:
# Grouping and transforming the user ID column
ords_prods['max_order'] = ords_prods.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/53/fhgd2kgx51j3l2vlj5dchl_w0000gq/T/ipykernel_99407/3861047130.py:2: FutureWarning: The provided callable <function max at 0x10ca09440> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods['max_order'] = ords_prods.groupby(['user_id'])['order_number'].transform(np.max)


In [31]:
# Checking the validity of the results
ords_prods.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices _merge    price_range_loc    busiest_days  \
0              7     9.0   both  Mid-range product  Regularly busy   
1             16    12.5   both  Mid-range product  Regularly busy   
2             19     4.4   both  Low-range product  Regularly busy   
3             19     4.7   both  Low-range product  Regularly busy   
4             17     1.0   both  Low-range product  Regularly busy   

  busiest_period_of_day  max_order  
0        Average Orders         10  
1        Average Orders         10  
2        Average Orders         10  
3        Average Orders         10  
4        Average Orders         10

In [34]:
# Getting rid of row limits

pd.options.display.max_rows = None

In [37]:
# Displaying the 100 rows

ords_prods.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2539329        1    prior             1                   2   
2    2539329        1    prior             1                   2   
3    2539329        1    prior             1                   2   
4    2539329        1    prior             1                   2   
5    2398795        1    prior             2                   3   
6    2398795        1    prior             2                   3   
7    2398795        1    prior             2                   3   
8    2398795        1    prior             2                   3   
9    2398795        1    prior             2                   3   
10   2398795        1    prior             2                   3   
11    473747        1    prior             3                   3   
12    473747        1    prior             3                   3   
13    473747        1    prior             3                   3   
14    473747        1    prior             3                   3   
15    473747        1    prior             3                   3   
16   2254736        1    prior             4                   4   
17   2254736        1    prior             4                   4   
18   2254736        1    prior             4                   4   
19   2254736        1    prior             4                   4   
20   2254736        1    prior             4                   4   
21    431534        1    prior             5                   4   
22    431534        1    prior             5                   4   
23    431534        1    prior             5                   4   
24    431534        1    prior             5                   4   
25    431534        1    prior             5                   4   
26    431534        1    prior             5                   4   
27    431534        1    prior             5                   4   
28    431534        1    prior             5                   4   
29   3367565        1    prior             6                   2   
30   3367565        1    prior             6                   2   
31   3367565        1    prior             6                   2   
32   3367565        1    prior             6                   2   
33    550135        1    prior             7                   1   
34    550135        1    prior             7                   1   
35    550135        1    prior             7                   1   
36    550135        1    prior             7                   1   
37    550135        1    prior             7                   1   
38   3108588        1    prior             8                   1   
39   3108588        1    prior             8                   1   
40   3108588        1    prior             8                   1   
41   3108588        1    prior             8                   1   
42   3108588        1    prior             8                   1   
43   3108588        1    prior             8                   1   
44   2295261        1    prior             9                   1   
45   2295261        1    prior             9                   1   
46   2295261        1    prior             9                   1   
47   2295261        1    prior             9                   1   
48   2295261        1    prior             9                   1   
49   2295261        1    prior             9                   1   
50   2550362        1    prior            10                   4   
51   2550362        1    prior            10                   4   
52   2550362        1    prior            10                   4   
53   2550362        1    prior            10                   4   
54   2550362        1    prior            10                   4   
55   2550362        1    prior            10                   4   
56   2550362        1    prior            10                   4   
57   2550362        1    prior            10                

In [40]:
# Setting up the conditions for the new Loyalty column 

ords_prods.loc[ords_prods['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [43]:
ords_prods.loc[(ords_prods['max_order'] <= 40) & (ords_prods['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [46]:
ords_prods.loc[ords_prods['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [49]:
# Checking if the new column is correct

ords_prods['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [52]:
# Checking if everything is correct once more

ords_prods[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
5         1      New customer             2
6         1      New customer             2
7         1      New customer             2
8         1      New customer             2
9         1      New customer             2
10        1      New customer             2
11        1      New customer             3
12        1      New customer             3
13        1      New customer             3
14        1      New customer             3
15        1      New customer             3
16        1      New customer             4
17        1      New customer             4
18        1      New customer             4
19        1      New customer             4
20        1      New customer             4
21        1      New customer             5
22        1      New customer             5
23        1      New customer             5
24        1      New customer             5
25        1      New customer             5
26        1      New customer             5
27        1      New customer             5
28        1      New customer             5
29        1      New customer             6
30        1      New customer             6
31        1      New customer             6
32        1      New customer             6
33        1      New customer             7
34        1      New customer             7
35        1      New customer             7
36        1      New customer             7
37        1      New customer             7
38        1      New customer             8
39        1      New customer             8
40        1      New customer             8
41        1      New customer             8
42        1      New customer             8
43        1      New customer             8
44        1      New customer             9
45        1      New customer             9
46        1      New customer             9
47        1      New customer             9
48        1      New customer             9
49        1      New customer             9
50        1      New customer            10
51        1      New customer            10
52        1      New customer            10
53        1      New customer            10
54        1      New customer            10
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1

# 05. Identifying spending habits by customer loyalty

In [63]:
# Grouping and aggregating the info based on customer loyalty

ords_prods.groupby('loyalty_flag').agg({'prices': ['min', 'max', 'mean']})

prices                
                    min   max      mean
loyalty_flag                           
Loyal customer      1.0  25.0  7.773487
New customer        1.0  25.0  7.801077
Regular customer    1.0  25.0  7.798145

There seems to be no real difference between the spending habits of customers with different loyalty flags.

# 06. Targeting high spenders

In [89]:
# Creating a new column for average spending

ords_prods['average_spending'] = ords_prods.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/53/fhgd2kgx51j3l2vlj5dchl_w0000gq/T/ipykernel_99407/1358155528.py:3: FutureWarning: The provided callable <function mean at 0x10ca09e40> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods['average_spending'] = ords_prods.groupby(['user_id'])['prices'].transform(np.mean)


In [93]:
ords_prods.head(60)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2539329        1    prior             1                   2   
2    2539329        1    prior             1                   2   
3    2539329        1    prior             1                   2   
4    2539329        1    prior             1                   2   
5    2398795        1    prior             2                   3   
6    2398795        1    prior             2                   3   
7    2398795        1    prior             2                   3   
8    2398795        1    prior             2                   3   
9    2398795        1    prior             2                   3   
10   2398795        1    prior             2                   3   
11    473747        1    prior             3                   3   
12    473747        1    prior             3                   3   
13    473747        1    prior             3                   3   
14    473747        1    prior             3                   3   
15    473747        1    prior             3                   3   
16   2254736        1    prior             4                   4   
17   2254736        1    prior             4                   4   
18   2254736        1    prior             4                   4   
19   2254736        1    prior             4                   4   
20   2254736        1    prior             4                   4   
21    431534        1    prior             5                   4   
22    431534        1    prior             5                   4   
23    431534        1    prior             5                   4   
24    431534        1    prior             5                   4   
25    431534        1    prior             5                   4   
26    431534        1    prior             5                   4   
27    431534        1    prior             5                   4   
28    431534        1    prior             5                   4   
29   3367565        1    prior             6                   2   
30   3367565        1    prior             6                   2   
31   3367565        1    prior             6                   2   
32   3367565        1    prior             6                   2   
33    550135        1    prior             7                   1   
34    550135        1    prior             7                   1   
35    550135        1    prior             7                   1   
36    550135        1    prior             7                   1   
37    550135        1    prior             7                   1   
38   3108588        1    prior             8                   1   
39   3108588        1    prior             8                   1   
40   3108588        1    prior             8                   1   
41   3108588        1    prior             8                   1   
42   3108588        1    prior             8                   1   
43   3108588        1    prior             8                   1   
44   2295261        1    prior             9                   1   
45   2295261        1    prior             9                   1   
46   2295261        1    prior             9                   1   
47   2295261        1    prior             9                   1   
48   2295261        1    prior             9                   1   
49   2295261        1    prior             9                   1   
50   2550362        1    prior            10                   4   
51   2550362        1    prior            10                   4   
52   2550362        1    prior            10                   4   
53   2550362        1    prior            10                   4   
54   2550362        1    prior            10                   4   
55   2550362        1    prior            10                   4   
56   2550362        1    prior            10                   4   
57   2550362        1    prior            10                

In [95]:
# Setting up a new column for the averages that will indicate the spending trends

ords_prods.loc[ords_prods['average_spending'] < 10, 'spending_flag'] = 'Low Spender'

In [97]:
ords_prods.loc[ords_prods['average_spending'] >= 10, 'spending_flag'] = 'High Spender'

In [103]:
# Checking the results

ords_prods['spending_flag'].value_counts(dropna = False)

spending_flag
Low Spender     32285150
High Spender      119709
Name: count, dtype: int64

In [109]:
# Checking results once more 

ords_prods[['user_id', 'spending_flag', 'average_spending']].head(100)

user_id spending_flag  average_spending
0         1   Low Spender          6.367797
1         1   Low Spender          6.367797
2         1   Low Spender          6.367797
3         1   Low Spender          6.367797
4         1   Low Spender          6.367797
5         1   Low Spender          6.367797
6         1   Low Spender          6.367797
7         1   Low Spender          6.367797
8         1   Low Spender          6.367797
9         1   Low Spender          6.367797
10        1   Low Spender          6.367797
11        1   Low Spender          6.367797
12        1   Low Spender          6.367797
13        1   Low Spender          6.367797
14        1   Low Spender          6.367797
15        1   Low Spender          6.367797
16        1   Low Spender          6.367797
17        1   Low Spender          6.367797
18        1   Low Spender          6.367797
19        1   Low Spender          6.367797
20        1   Low Spender          6.367797
21        1   Low Spender          6.367797
22        1   Low Spender          6.367797
23        1   Low Spender          6.367797
24        1   Low Spender          6.367797
25        1   Low Spender          6.367797
26        1   Low Spender          6.367797
27        1   Low Spender          6.367797
28        1   Low Spender          6.367797
29        1   Low Spender          6.367797
30        1   Low Spender          6.367797
31        1   Low Spender          6.367797
32        1   Low Spender          6.367797
33        1   Low Spender          6.367797
34        1   Low Spender          6.367797
35        1   Low Spender          6.367797
36        1   Low Spender          6.367797
37        1   Low Spender          6.367797
38        1   Low Spender          6.367797
39        1   Low Spender          6.367797
40        1   Low Spender          6.367797
41        1   Low Spender          6.367797
42        1   Low Spender          6.367797
43        1   Low Spender          6.367797
44        1   Low Spender          6.367797
45        1   Low Spender          6.367797
46        1   Low Spender          6.367797
47        1   Low Spender          6.367797
48        1   Low Spender          6.367797
49        1   Low Spender          6.367797
50        1   Low Spender          6.367797
51        1   Low Spender          6.367797
52        1   Low Spender          6.367797
53        1   Low Spender          6.367797
54        1   Low Spender          6.367797
55        1   Low Spender          6.367797
56        1   Low Spender          6.367797
57        1   Low Spender          6.367797
58        1   Low Spender          6.367797
59        2   Low Spender          7.515897
60        2   Low Spender          7.515897
61        2   Low Spender          7.515897
62        2   Low Spender          7.515897
63        2   Low Spender          7.515897
64        2   Low Spender          7.515897
65        2   Low Spender          7.515897
66        2   Low Spender          7.515897
67        2   Low Spender          7.515897
68        2   Low Spender          7.515897
69        2   Low Spender          7.515897
70        2   Low Spender          7.515897
71        2   Low Spender          7.515897
72        2   Low Spender          7.515897
73        2   Low Spender          7.515897
74        2   Low Spender          7.515897
75        2   Low Spender          7.515897
76        2   Low Spender          7.515897
77        2   Low Spender          7.515897
78        2   Low Spender          7.515897
79        2   Low Spender          7.515897
80        2   Low Spender          7.515897
81        2   Low Spender          7.515897
82        2   Low Spender          7.515897
83        2   Low Spender          7.515897
84        2   Low Spender          7.515897
85        2   Low Spender          7.515897
86        2   Low Spender          7.515897
87        2   Low Spender          7.515897
88        2   Low Spender          7.515897
89        2   Low Spender          7.515897


# 07. Identifying the frequent customers

In [113]:
# Creating a new column for buying frequency median

ords_prods['buying_frequency'] = ords_prods.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/53/fhgd2kgx51j3l2vlj5dchl_w0000gq/T/ipykernel_99407/4078048832.py:3: FutureWarning: The provided callable <function median at 0x10cbed4e0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods['buying_frequency'] = ords_prods.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [119]:
# Checking the results

ords_prods.head(60)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2539329        1    prior             1                   2   
2    2539329        1    prior             1                   2   
3    2539329        1    prior             1                   2   
4    2539329        1    prior             1                   2   
5    2398795        1    prior             2                   3   
6    2398795        1    prior             2                   3   
7    2398795        1    prior             2                   3   
8    2398795        1    prior             2                   3   
9    2398795        1    prior             2                   3   
10   2398795        1    prior             2                   3   
11    473747        1    prior             3                   3   
12    473747        1    prior             3                   3   
13    473747        1    prior             3                   3   
14    473747        1    prior             3                   3   
15    473747        1    prior             3                   3   
16   2254736        1    prior             4                   4   
17   2254736        1    prior             4                   4   
18   2254736        1    prior             4                   4   
19   2254736        1    prior             4                   4   
20   2254736        1    prior             4                   4   
21    431534        1    prior             5                   4   
22    431534        1    prior             5                   4   
23    431534        1    prior             5                   4   
24    431534        1    prior             5                   4   
25    431534        1    prior             5                   4   
26    431534        1    prior             5                   4   
27    431534        1    prior             5                   4   
28    431534        1    prior             5                   4   
29   3367565        1    prior             6                   2   
30   3367565        1    prior             6                   2   
31   3367565        1    prior             6                   2   
32   3367565        1    prior             6                   2   
33    550135        1    prior             7                   1   
34    550135        1    prior             7                   1   
35    550135        1    prior             7                   1   
36    550135        1    prior             7                   1   
37    550135        1    prior             7                   1   
38   3108588        1    prior             8                   1   
39   3108588        1    prior             8                   1   
40   3108588        1    prior             8                   1   
41   3108588        1    prior             8                   1   
42   3108588        1    prior             8                   1   
43   3108588        1    prior             8                   1   
44   2295261        1    prior             9                   1   
45   2295261        1    prior             9                   1   
46   2295261        1    prior             9                   1   
47   2295261        1    prior             9                   1   
48   2295261        1    prior             9                   1   
49   2295261        1    prior             9                   1   
50   2550362        1    prior            10                   4   
51   2550362        1    prior            10                   4   
52   2550362        1    prior            10                   4   
53   2550362        1    prior            10                   4   
54   2550362        1    prior            10                   4   
55   2550362        1    prior            10                   4   
56   2550362        1    prior            10                   4   
57   2550362        1    prior            10                

In [121]:
# Creating a new column for frequency flag

ords_prods.loc[ords_prods['buying_frequency'] > 20, 'frequency_flag'] = 'Non-Frequent Customer'

In [136]:
ords_prods.loc[(ords_prods['buying_frequency'] <= 20) & (ords_prods['buying_frequency'] > 10), 'frequency_flag'] = 'Regular Customer'

In [140]:
ords_prods.loc[ords_prods['buying_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [148]:
# Checking the results

ords_prods['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        21559853
Regular Customer          7208564
Non-Frequent Customer     3636437
NaN                             5
Name: count, dtype: int64

In [150]:
# Checking results once more 

ords_prods[['user_id', 'frequency_flag', 'buying_frequency']].head(60)

user_id         frequency_flag  buying_frequency
0         1  Non-Frequent Customer              20.5
1         1  Non-Frequent Customer              20.5
2         1  Non-Frequent Customer              20.5
3         1  Non-Frequent Customer              20.5
4         1  Non-Frequent Customer              20.5
5         1  Non-Frequent Customer              20.5
6         1  Non-Frequent Customer              20.5
7         1  Non-Frequent Customer              20.5
8         1  Non-Frequent Customer              20.5
9         1  Non-Frequent Customer              20.5
10        1  Non-Frequent Customer              20.5
11        1  Non-Frequent Customer              20.5
12        1  Non-Frequent Customer              20.5
13        1  Non-Frequent Customer              20.5
14        1  Non-Frequent Customer              20.5
15        1  Non-Frequent Customer              20.5
16        1  Non-Frequent Customer              20.5
17        1  Non-Frequent Customer              20.5
18        1  Non-Frequent Customer              20.5
19        1  Non-Frequent Customer              20.5
20        1  Non-Frequent Customer              20.5
21        1  Non-Frequent Customer              20.5
22        1  Non-Frequent Customer              20.5
23        1  Non-Frequent Customer              20.5
24        1  Non-Frequent Customer              20.5
25        1  Non-Frequent Customer              20.5
26        1  Non-Frequent Customer              20.5
27        1  Non-Frequent Customer              20.5
28        1  Non-Frequent Customer              20.5
29        1  Non-Frequent Customer              20.5
30        1  Non-Frequent Customer              20.5
31        1  Non-Frequent Customer              20.5
32        1  Non-Frequent Customer              20.5
33        1  Non-Frequent Customer              20.5
34        1  Non-Frequent Customer              20.5
35        1  Non-Frequent Customer              20.5
36        1  Non-Frequent Customer              20.5
37        1  Non-Frequent Customer              20.5
38        1  Non-Frequent Customer              20.5
39        1  Non-Frequent Customer              20.5
40        1  Non-Frequent Customer              20.5
41        1  Non-Frequent Customer              20.5
42        1  Non-Frequent Customer              20.5
43        1  Non-Frequent Customer              20.5
44        1  Non-Frequent Customer              20.5
45        1  Non-Frequent Customer              20.5
46        1  Non-Frequent Customer              20.5
47        1  Non-Frequent Customer              20.5
48        1  Non-Frequent Customer              20.5
49        1  Non-Frequent Customer              20.5
50        1  Non-Frequent Customer              20.5
51        1  Non-Frequent Customer              20.5
52        1  Non-Frequent Customer              20.5
53        1  Non-Frequent Customer              20.5
54        1  Non-Frequent Customer              20.5
55        1  Non-Frequent Customer              20.5
56        1  Non-Frequent Customer              20.5
57        1  Non-Frequent Customer              20.5
58        1  Non-Frequent Customer              20.5
59        2       Regular Customer              13.0

In [153]:
# Exporting the file

ords_prods.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_agg.pkl'))